src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -45,63 +50,74 @@ import           ZkFold.Base.Algebra.Polynomials.Multivariate (Mono, Poly, evalM
 import           ZkFold.Base.Control.HApplicative
 import           ZkFold.Base.Data.HFunctor
 import           ZkFold.Base.Data.Package
-import           ZkFold.Prelude                               (drop, length)
 import           ZkFold.Symbolic.Class
 import           ZkFold.Symbolic.MonadCircuit
 
 -- | Arithmetic circuit in the form of a system of polynomial constraints.
-data ArithmeticCircuit a o = ArithmeticCircuit
+data ArithmeticCircuit a i o = ArithmeticCircuit
     {
-        acSystem   :: Map Natural (Constraint a),
+        acSystem   :: Map Natural (Constraint a i),
         -- ^ The system of polynomial constraints
         acRange    :: Map Natural a,
         -- ^ The range constraints [0, a] for the selected variables
-        acInput    :: [Natural],
-        -- ^ The input variables
-        acWitness  :: Map Natural (Map Natural a -> a),
+        acWitness  :: Map Natural (i a -> a),
         -- ^ The witness generation functions
         acVarOrder :: Map (Natural, Natural) Natural,
         -- ^ The order of variable assignments
         acRNG      :: StdGen,
         -- ^ random generator for generating unique variables
-        acOutput   :: o Natural
+        acOutput   :: o (Var i)
         -- ^ The output variables
     } deriving (Generic)
-deriving instance (NFData a, NFData (o Natural))
-    => NFData (ArithmeticCircuit a o)
-
-witnessGenerator :: ArithmeticCircuit a o -> Map Natural a -> Map Natural a
+deriving instance (NFData a, NFData (o (Var i)), NFData (Rep i))
+    => NFData (ArithmeticCircuit a i o)
+
+acInput :: Representable i => i (Var i)
+acInput = fmapRep InVar (tabulate id)
+
+data Var i
+  = InVar (Rep i)
+  | NewVar Natural
+  deriving Generic
+deriving anyclass instance FromJSON (Rep i) => FromJSON (Var i)
+deriving anyclass instance FromJSON (Rep i) => FromJSONKey (Var i)
+deriving anyclass instance ToJSON (Rep i) => ToJSONKey (Var i)
+deriving anyclass instance ToJSON (Rep i) => ToJSON (Var i)
+deriving stock instance Show (Rep i) => Show (Var i)
+deriving stock instance Eq (Rep i) => Eq (Var i)
+deriving stock instance Ord (Rep i) => Ord (Var i)
+deriving instance NFData (Rep i) => NFData (Var i)
+
+witnessGenerator :: ArithmeticCircuit a i o -> i a -> Map Natural a
 witnessGenerator circuit inputs =
-  let srcs = acWitness circuit
-      witness = ($ witness) <$> (srcs `union` fmap const inputs)
-   in witness
+    fmap ($ inputs) (acWitness circuit)
```

Comment:
Because of this simpler setup, witness computation takes (exponentially?) more time because now each new value in a witness recomputes all other values it depends on, they do the same...
The same loeb trick as was here previously is achievable with `acWitness :: Map Natural (i a -> Map Natural a -> a)`.